server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2216,10 +2218,11 @@ public void acquireReplicaOperationPermit(final long operationPrimaryTerm, final
                                               final Object debugInfo) {
         verifyNotClosed();
         verifyReplicationTarget();
-        final boolean globalCheckpointUpdated;
         if (operationPrimaryTerm > primaryTerm) {
             synchronized (primaryTermMutex) {
```

Comment:
I wonder if we can remove this and only lock `mutex` on this level (it's always good to avoid multiple locks if possible).

Reply:
I'm fine relaxing it and using mutex as we now use asyncBlockOperations. This means that in practice, we will have at most have the number of indexing threads block on this (while possibly a concurrent cluster state update comes in, trying to acquire mutex as well). The first indexing thread will increase `pendingPrimaryTerm`, and all the other ones that are blocked on mutex will just acquire mutex and do a quick noop. All subsequent writes will not acquire the mutex anymore as they will bypass the pre-flight check.